kernel/pci/src/lib.rs

```diff
@@ -217,78 +256,110 @@ impl PciLocation {
     pub fn function(&self) -> u8 { self.func }
 
 
-    /// Computes a PCI address from bus, slot, func, and offset. 
+    /// Computes a PCI address and value shift from bus, slot, func, and offset. 
     /// The least two significant bits of offset are masked, so it's 4-byte aligned addressing,
     /// which makes sense since we read PCI data (from the configuration space) in 32-bit chunks.
-    fn pci_address(self, offset: u8) -> u32 {
-        ((self.bus  as u32) << 16) | 
-        ((self.slot as u32) << 11) | 
-        ((self.func as u32) <<  8) | 
-        ((offset as u32) & (PCI_CONFIG_ADDRESS_OFFSET_MASK as u32)) | 
-        0x8000_0000
+    /// The shift allows selecting a byte/word from the addressed value.
+    fn pci_address_and_shift(self, offset: u8) -> (u32, u32) {
```

Comment:
more docs would be welcome here, especially for how `shift` works, e.g., it's used to right-shift the PCI data value from pci_read_32

It'd also be a good idea to use a struct for the return type, and/or a newtype wrapper for the address and shift value too (e.g., `PciAddress` and `PciShift`)

Reply:
New doc and struct:
```rust
struct PciAddress {
    /// Offset to a double word in a PCI configuration space
    dword_address: u32,
    /// Bit offset of the bytes of interest in this dword
    byte_shift: u32,
}

    /// Computes a [`PciAddress`] from bus, slot, func, and offset. 
    ///
    /// PCI configuration space addresses must be aligned to 4 bytes (double word)
    /// for reads and writes to succeed. The `dword_address` field of the returned
    /// structure will be aligned to 4 bytes accordingly.
    ///
    /// Offsets with the two least significant bits set are still valid (they point
    /// to words and bytes). This function computes a `byte_shift` value which should
    /// be used to shift the double word pointed to by `dword_address`.
```